### ISSUE TITLE
Skip non-Tensor/Module frames

### ISSUE THREAD

**Issue by jansel** [pytorch/torchdynamo#69]
People may have non-PyTorch code run under TorchDynamo.  We should make sure TorchDynamo does nothing in this case.

If TorchDynamo reaches the end of the frame without finding PyTorch ops, it will [just run the frame normally](https://github.com/facebookresearch/torchdynamo/blob/44971ffd9a7e6798b7868a592c337acb75bd1d2d/torchdynamo/symbolic_convert.py#L1056).  However, if there are unsupported things that prevent a whole-graph, TorchDyanmo could generate specialized frames for non-PyTorch code.   This should be correct, but could add extra overhead.

To improve this we should expand the logic in this function:
https://github.com/facebookresearch/torchdynamo/blob/44971ffd9a7e6798b7868a592c337acb75bd1d2d/torchdynamo/symbolic_convert.py#L1008
That function controls if TorchDynamo should break the graph and generate a `resume_at_xx` function to pick up after an unsupported thing.

The logic I would propose is: examine the stack, locals, and globals referenced by co_names; if there is a tensor/nn.Module/torch.* anywhere then, then keep doing what we do now; if there is not, just bail out and switch to normal execution.

**Comment by jansel** [comment]
> The logic I would propose is: examine the stack, locals, and globals referenced by co_names; if there is a tensor/nn.Module/torch.* anywhere then, then keep doing what we do now; if there is not, just bail out and switch to normal execution.

On further discussion with folks such as @ezyang.  I think we should run this logic at the start of each frame, before we start doing any work to try to reduce overheads of symbolic analysis.

**Comment by ezyang** [comment]
A very minor problem with the heuristics we know how to do is that this won't work:

```
def factory_wrapper(args):
    return torch.empty(args)

def foo(args):
    x = factory_wrapper(args)
    return do_something(x)
```

where `do_something` contains Torch API calls; we won't start analysis on foo. But this seems acceptable (there is no way to determine that we should trace this locally), and maybe if the user adds some type annotations we can pick up on this case too.

**Comment by jansel** [comment]
We should make the decorator syntax:
```
@torchdynamo.optimize(...)
def foo(args):
   ...
```
force `foo.__code__` to pass the checks.  

That way users have a way around a false-fast-exit.

**Comment by anijain2305** [comment]
Happy to look into this. Assigning to myself. Let me know if someone is already looking into this.

**Comment by ezyang** [comment]
@anijain2305 this will be super useful for broadening torchdynamo's testing efforts; right now running torchdynamo on something like PyTorch's test suite ends with torchdynamo rabbitholing into a bunch of random crap which could have been skipped with the functionality here.

If you want a torture test, run this test in Python:

```
python test/test_torch.py test/test_torch.py TestTorch.test_RNG_after_pickle
```

with dynamo enabled with this patch:

```
diff --git a/torch/testing/_internal/common_utils.py b/torch/testing/_internal/common_utils.py
index 0b072919c4..549c1e9942 100644
--- a/torch/testing/_internal/common_utils.py
+++ b/torch/testing/_internal/common_utils.py
@@ -1794,7 +1794,11 @@ class TestCase(expecttest.TestCase):
             failures_before = 0 if result is None else len(result.failures)  # num tests marked as failed before starting
             errors_before = 0 if result is None else len(result.errors)  # num tests marked as errored before starting
 
-        super().run(result=result)
+        import torchdynamo
+        torchdynamo.config.trace = True
+        with torchdynamo.optimize("eager"):
+            super().run(result=result)
+
         # Early terminate test if necessary.
         if self._should_stop_test_suite():
             if result.wasSuccessful():
```

**Comment by ezyang** [comment]
(you should also revert my bandaid PR https://github.com/facebookresearch/torchdynamo/pull/157 )

**Comment by anijain2305** [comment]
Thanks @ezyang Thats a good starting point. Will look into this.

**Comment by anijain2305** [comment]
@jansel 

A large number of errors are because of torch.* ops that do not return Tensor objects. In the past, I added them one by one for TorchBench models. But with `test_torch.py`, there are just too many (pasted below). One quick way is to just use `unimplemented` instead of assert and skip rest of the frame, but it can leave performance on the table. Do you have any suggestion?

~~~
AssertionError: torch.* op returned non-Tensor dtype call_function <built-in function getitem>
AssertionError: torch.* op returned non-Tensor int call_function <function _element_size at 0x7f8706341160>
AssertionError: torch.* op returned non-Tensor dtype call_function <built-in function getitem>
AssertionError: torch.* op returned non-Tensor dtype call_function <built-in function getitem>
AssertionError: torch.* op returned non-Tensor bool call_function <built-in method can_cast of type object at 0x7f86da5ef720>
AssertionError: torch.* op returned non-Tensor dtype call_function <built-in function getitem>
AssertionError: torch.* op returned non-Tensor product call_function <class 'itertools.product'>
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor _TypedStorage call_method storage
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor PyCapsule call_function <built-in function _to_dlpack>
AssertionError: torch.* op returned non-Tensor PyCapsule call_function <built-in function _to_dlpack>
AssertionError: torch.* op returned non-Tensor DeterministicGuard call_function <class 'torch.testing._internal.common_utils.DeterministicGuard'>
AssertionError: torch.* op returned non-Tensor DeterministicGuard call_function <class 'torch.testing._internal.common_utils.DeterministicGuard'>
AssertionError: torch.* op returned non-Tensor DeterministicGuard call_function <class 'torch.testing._internal.common_utils.DeterministicGuard'>
AssertionError: torch.* op returned non-Tensor DeterministicGuard call_function <class 'torch.testing._internal.common_utils.DeterministicGuard'>
AssertionError: torch.* op returned non-Tensor product call_function <class 'itertools.product'>
AssertionError: torch.* op returned non-Tensor DeterministicGuard call_function <class 'torch.testing._internal.common_utils.DeterministicGuard'>
AssertionError: torch.* op returned non-Tensor product call_function <class 'itertools.product'>
AssertionError: torch.* op returned non-Tensor bool call_method is_complex
AssertionError: torch.* op returned non-Tensor DeterministicGuard call_function <class 'torch.testing._internal.common_utils.DeterministicGuard'>
AssertionError: torch.* op returned non-Tensor product call_function <class 'itertools.product'>
AssertionError: torch.* op returned non-Tensor bool call_method is_set_to
AssertionError: torch.* op returned non-Tensor bool call_method is_signed
AssertionError: torch.* op returned non-Tensor product call_function <class 'itertools.product'>
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor expectedAlertNondeterministic call_function <class 'torch.testing._internal.common_device_type.expectedAlertNondeterministic'>
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.device'>
AssertionError: torch.* op returned non-Tensor DeterministicGuard call_function <class 'torch.testing._internal.common_utils.DeterministicGuard'>
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.device'>
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.device'>
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.device'>
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.device'>
AssertionError: torch.* op returned non-Tensor _TypedStorage call_method storage
AssertionError: torch.* op returned non-Tensor _TypedStorage call_method storage
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.device'>
AssertionError: torch.* op returned non-Tensor _TypedStorage call_method storage
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.device'>
AssertionError: torch.* op returned non-Tensor dtype call_function <built-in function getitem>
AssertionError: torch.* op returned non-Tensor dtype call_function <built-in function get_default_dtype>
AssertionError: torch.* op returned non-Tensor dtype call_function <built-in function get_default_dtype>
AssertionError: torch.* op returned non-Tensor bool call_function <built-in function not_>
AssertionError: torch.* op returned non-Tensor Stream call_function <class 'torch.cuda.streams.Stream'>
AssertionError: torch.* op returned non-Tensor Stream call_function <class 'torch.cuda.streams.Stream'>
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor Generator call_function <class 'torch._C.Generator'>
AssertionError: torch.* op returned non-Tensor bool call_method is_pinned
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.cuda.device'>
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.cuda.device'>
AssertionError: torch.* op returned non-Tensor _TypedStorage call_method storage
AssertionError: torch.* op returned non-Tensor CudaSyncGuard call_function <class 'torch.testing._internal.common_utils.CudaSyncGuard'>
AssertionError: torch.* op returned non-Tensor CudaSyncGuard call_function <class 'torch.testing._internal.common_utils.CudaSyncGuard'>
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.device'>
AssertionError: torch.* op returned non-Tensor product call_function <class 'itertools.product'>
AssertionError: torch.* op returned non-Tensor Generator call_function <class 'torch._C.Generator'>
AssertionError: torch.* op returned non-Tensor int call_function <built-in function get_num_threads>
AssertionError: torch.* op returned non-Tensor bool call_method __contains__
AssertionError: torch.* op returned non-Tensor dtype call_function <built-in function getitem>
AssertionError: torch.* op returned non-Tensor str call_function <function _cxx_flags at 0x7f869dd1c0d0>
AssertionError: torch.* op returned non-Tensor product call_function <class 'itertools.product'>
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.device'>
AssertionError: torch.* op returned non-Tensor dtype call_function <built-in function getitem>
AssertionError: torch.* op returned non-Tensor _TypedStorage call_function <class 'torch.ByteStorage'>
AssertionError: torch.* op returned non-Tensor bool call_method equal
AssertionError: torch.* op returned non-Tensor _GeneratorContextManager call_function <function TemporaryFileName at 0x7f869db9cca0>
AssertionError: torch.* op returned non-Tensor Generator call_function <class 'torch._C.Generator'>
AssertionError: torch.* op returned non-Tensor int call_function <built-in method _debug_has_internal_overlap of type object at 0x7f86da5ef720>
AssertionError: torch.* op returned non-Tensor _TypedStorage call_method storage
AssertionError: torch.* op returned non-Tensor str call_function <built-in function getitem>
AssertionError: torch.* op returned non-Tensor str call_function <built-in function getitem>
AssertionError: torch.* op returned non-Tensor bool call_method is_same_size
AssertionError: torch.* op returned non-Tensor int call_function <function initial_seed at 0x7f869e89f550>
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
~~~

**Comment by jansel** [comment]
@anijain2305 the root issue here is in [allowed_functions.py](https://github.com/facebookresearch/torchdynamo/blob/main/torchdynamo/allowed_functions.py).

TorchDynamo expects everything marked by allowed_functions to be a PyTorch operator that can be placed in an FX graph.  It generates allowed_functions by walking `torch.*`.  Most of these things should just be removed from allowed_functions, and perhaps we want to rethink how allowed_functions is generated.

Using unimplemented() would be an ok workaround, though it might be good to try to work through that list of functions and do something smarter.  

E.g. for things like `is_complex()` `is_contiguous()` we might bee able to just return constants similar to how `.size()` or `.dtype` is handled today.

**Comment by jansel** [comment]
Specifically, I'd expect these things to be pretty easy to handle (most just returning constants we guard on):
```
AssertionError: torch.* op returned non-Tensor bool call_function <built-in method can_cast of type object at 0x7f86da5ef720>
AssertionError: torch.* op returned non-Tensor bool call_method is_complex
AssertionError: torch.* op returned non-Tensor bool call_method is_contiguous
AssertionError: torch.* op returned non-Tensor bool call_method is_pinned
AssertionError: torch.* op returned non-Tensor bool call_method is_same_size
AssertionError: torch.* op returned non-Tensor bool call_method is_signed
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.cuda.device'>
AssertionError: torch.* op returned non-Tensor device call_function <class 'torch.device'>
AssertionError: torch.* op returned non-Tensor dtype call_function <built-in function get_default_dtype>
AssertionError: torch.* op returned non-Tensor int call_function <function _element_size at 0x7f8706341160>
AssertionError: torch.* op returned non-Tensor product call_function <class 'itertools.product'>
```

**Comment by ezyang** [comment]
ops are derived mechanically from `native_functions.yaml` so if you need us to generate some metadata to populate some tables we can totally do this (or you could use hot off the presses torchgen to work it out at runtime).

**Comment by jansel** [comment]
> ops are derived mechanically from `native_functions.yaml` so if you need us to generate some metadata to populate some tables we can totally do this (or you could use hot off the presses torchgen to work it out at runtime).

Wouldn't that be just for aten ops, or does it have Python API ops too?

**Comment by ezyang** [comment]
oh you know what, I dun' goofed, I read it as "torch.ops" and assumed we were talking about torch.ops.aten. My bad.

**Comment by anijain2305** [comment]
## Current status of test_torch  

There are 2 categories

1. Test failures
2. Test passes but TorchDynamo raises InternalTorchDynamo error and prints the stack trace.

Currently, I am focusing on the first category.

We started the week with 167 failures. And today we are at 53 failures. The fixes are #174, #182 and #184

------

## Notes on the remaining failures
In the remaining 53 failures, there are 2 categories which will bring the number significantly
* CUDA memory leak - 
~~~
RuntimeError: CUDA driver API confirmed a leak in test_torch.TestTorchDeviceTypeCUDA.test_large_cumprod_cuda_float16! Caching allocator allocated memory was 859136 and is now reported as 2148343296 on device 0. CUDA driver allocated memory was 1822425088 and is now 39720
~~~

* Tensor references - Tests like the following are failing
~~~
    def test_dead_weak_ref(self):
        x = torch.empty(2)
        w_x = weakref.ref(x)
        y = torch.empty(2)
        y.grad = x
        del x

        x = w_x()
        # Ideally, x would keep the tensor live.  But CPython doesn't
        # provide enough hooks to do this.  So it will go dead and x
        # will transmute into an undefined tensor.  Not great, but the
        # best we can do.
        del y

        self.assertRaises(RuntimeError, lambda: x.sigmoid())
~~~

**Comment by ezyang** [comment]
This is great work. I will be very interested to see what the root cause of CUDA memory leak and tensor references are. I wonder if dynamo is holding on to more references than it should; if so, it's very good that we will have caught this early :>

**Comment by anijain2305** [comment]
A quick update. For `test_torch.py`, we have resolved all memory leaks, yayy! And the current failure count stands at 15 (last time it was 53 failures).

There are some tests related to reference cycles that are not applicable to Dynamo. For example, in case of continuations, Dynamo has a chain of stack frames that might keep an object alive until all the stack frames have finished execution. Some tests assume that `del` will actually delete the tensor, but that might happen later in the case of TorchDynamo. I will find and skip such tests for TorchDynamo.


More broadly, I tried running TorchDynamo on all the PyTorch tests
~~~
pytest -v -s -n 32 test/

= 1588 failed, 31689 passed, 3579 skipped, 398 xfailed, 6396 warnings, 338 errors in 2215.56s (0:36:55) =
~~~

I am gonna start trying to get that number down now.

**Comment by jansel** [comment]
Wohoo!  Making some progress, thanks for grinding down those errors @anijain2305!

cc @soumith @ezyang for visibility

**Comment by ezyang** [comment]
> There are some tests related to reference cycles that are not applicable to Dynamo. For example, in case of continuations, Dynamo has a chain of stack frames that might keep an object alive until all the stack frames have finished execution.

Can you share some of these? Quite curious about it. Typically, the assumption that `del`  frees a tensor is usually for a tensor that has no refcount except in the current stack frame, and we're del'ing it to get rid of that sole reference. If Dynamo is keeping a chain of stack frames, I'd still expect it to free.

**Comment by anijain2305** [comment]
For example, this one https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L8253-L8265

~~~
    def test_dead_weak_ref(self):
        x = torch.empty(2)
        w_x = weakref.ref(x)
        y = torch.empty(2)
        y.grad = x
        del x

        x = w_x()
        # Ideally, x would keep the tensor live.  But CPython doesn't
        # provide enough hooks to do this.  So it will go dead and x
        # will transmute into an undefined tensor.  Not great, but the
        # best we can do.
        del y
        
        self.assertRaises(RuntimeError, lambda: x.sigmoid())
~~~

`weakref.ref(x)`, `y.grad = x` and `x = w_x()` result in graph break. TorchDynamo creates many frames with continuations here. So, now where we get to the frame that calls `self.assertRaises(RuntimeError, lambda: x.sigmoid())`, the `x` is alive in the outer frames. Thats my current understanding.

**Comment by ezyang** [comment]
OK, this makes sense, but I'd argue it's still a torchdynamo bug: instead of replicating x in the new frames, the accesses to x should be modeled as nonlocal accesses to a single occurrence of x in the parent frame. I don't know if Python bytecode lets you actually do this.

**Comment by ezyang** [comment]
FWIW, I think it's worth fixing this, because people DO expect prompt deallocation when they del, and if a graph break can cause this to no longer hold it will be very surprising.

**Comment by jansel** [comment]
It would be possible to change the way we codegen continuations to create a closure and replace all LOAD_FAST with LOAD_DEREF.  This would centralize all variable handling like you described, at the cost of an extra pointer deref on every access.

Another option would be to return from the original frame (releasing the reference counts), then have some code in C to call the continuation.

Both of those are totally doable, but would require extra bytecode analysis and rewriting.  The current simple "jump into the original bytecode" wouldn't work.   

I agree with @ezyang it is worth fixing (even though technically the python spec makes few guarantees about garbage collection semantics), though I would assign lower priority to it because I don't think it is a common case.

In the common case TorchDynamo (sans bugs and leaks) should use less memory than eager and be more aggressive about freeing unused memory.  Consider this example:
```
def foo(x):
   a=x+1
   b=a+1
   c=b+1
   return c+1  # In eager a/b/c get freed at the return
```

TorchDynamo will capture this entire thing in an FX graph, and FX will codegen look something like this:
```
def fx_graph(x):
   a=x+1; x=None
   b=a+1; a=None  # a freed
   c=b+1; b=None  # b freed
   return (c+1,)  # c freed
```
Which will lower peak memory usage for most models.

The issue described only happens if:
1) you allocate a var
2) you have a graph break
3) you `del` that var

It could happen.  Though I think on net TorchDynamo will improve memory usage more often than it hurts it.  (After we fix the remaining issues with symbolic propagation and leaks.)

**Comment by anijain2305** [comment]
cc @mlazos This is the parent thread. The instructions to run PyTorch tests are below

* Add this patch to pytorch

~~~
diff --git a/torch/testing/_internal/common_utils.py b/torch/testing/_internal/common_utils.py
index e1417ed1dc..abade35a8d 100644
--- a/torch/testing/_internal/common_utils.py
+++ b/torch/testing/_internal/common_utils.py
@@ -1826,7 +1826,21 @@ class TestCase(expecttest.TestCase):
             failures_before = 0 if result is None else len(result.failures)  # num tests marked as failed before starting
             errors_before = 0 if result is None else len(result.errors)  # num tests marked as errored before starting

-        super().run(result=result)
+        use_dynamo = True
+
+        if use_dynamo:
+            import torchdynamo
+            # torchdynamo.config.trace = True
+            # torchdynamo.config.debug = True
+            torchdynamo.reset()
+            with torchdynamo.optimize("eager"):
+                super().run(result=result)
+
+            # torchdynamo.reset()
+            # gc.collect()
+        else:
+            super().run(result=result)
+
         # Early terminate test if necessary.
         if self._should_stop_test_suite():
             if result.wasSuccessful():
~~~


* And run this test

~~~
CUDA_VISIBLE_DEVICES=-1 pytest -v -s -n 96 test/test_nn.py
~~~

Currently, there are 34 failures

~~~
FAILED test/test_nn.py::TestLazyModules::test_lazy_batchnorm1d - AssertionErr...
FAILED test/test_nn.py::TestLazyModules::test_lazy_batchnorm1d_pickle - Asser...
FAILED test/test_nn.py::TestLazyModules::test_lazy_batchnorm2d - AssertionErr...
FAILED test/test_nn.py::TestLazyModules::test_lazy_batchnorm2d_pickle - Asser...
FAILED test/test_nn.py::TestLazyModules::test_lazy_batchnorm3d - AssertionErr...
FAILED test/test_nn.py::TestLazyModules::test_lazy_batchnorm3d_pickle - Asser...
FAILED test/test_nn.py::TestLazyModules::test_lazy_instancenorm1d - Assertion...
FAILED test/test_nn.py::TestLazyModules::test_lazy_instancenorm1d_pickle - As...
FAILED test/test_nn.py::TestLazyModules::test_lazy_instancenorm2d - Assertion...
FAILED test/test_nn.py::TestLazyModules::test_lazy_instancenorm2d_pickle - As...
FAILED test/test_nn.py::TestLazyModules::test_lazy_instancenorm3d - Assertion...
FAILED test/test_nn.py::TestLazyModules::test_lazy_instancenorm3d_pickle - As...
FAILED test/test_nn.py::TestModuleGlobalHooks::test_global_and_local_hooks_order
FAILED test/test_nn.py::TestModuleGlobalHooks::test_module_forward_forward_hook_removable
FAILED test/test_nn.py::TestModuleGlobalHooks::test_module_forward_preforward_hook_removable
FAILED test/test_nn.py::TestModuleGlobalHooks::test_module_global_forward_preforward_hook_writeable
FAILED test/test_nn.py::TestModuleGlobalHooks::test_module_global_hooks - Ass...
FAILED test/test_nn.py::TestNN::test_ModuleDict - AssertionError: Linear(in_f...
FAILED test/test_nn.py::TestNN::test_ParameterDict - AssertionError: Paramete...
FAILED test/test_nn.py::TestNN::test_ParameterList - IndexError: index 1 is o...
FAILED test/test_nn.py::TestNN::test_ParameterList_replication - IndexError: ...
FAILED test/test_nn.py::TestNN::test_functional_grad_conv - AssertionError: S...
FAILED test/test_nn.py::TestNN::test_hook_backward_writeable - AssertionError...
FAILED test/test_nn.py::TestNN::test_hook_forward_preforward_writable - Asser...
FAILED test/test_nn.py::TestNN::test_hooks - AssertionError: Scalars are not ...
FAILED test/test_nn.py::TestNN::test_load_state_dict_ref_cycle - AssertionErr...
FAILED test/test_nn.py::TestNN::test_parameter_assignment - AssertionError: S...
FAILED test/test_nn.py::TestNN::test_parameterlistdict_pickle - IndexError: i...
FAILED test/test_nn.py::TestNN::test_parameterlistdict_setting_attributes - I...
FAILED test/test_nn.py::TestNN::test_share_memory - IndexError: index 1 is ou...
FAILED test/test_nn.py::TestNN::test_spectral_norm - torch.autograd.gradcheck...
FAILED test/test_nn.py::TestNNDeviceTypeCPU::test_clip_grad_norm_error_if_nonfinite_cpu
FAILED test/test_nn.py::TestNNDeviceTypeCPU::test_triplet_margin_with_distance_loss_cpu
FAILED test/test_nn.py::TestNNInit::test_dirac_identity - AssertionError: The...
~~~

**Comment by anijain2305** [comment]
@mlaos will be looking at the errors from three files in the start
* test_ops.py
* test_nn.py
* test_autograd.py

I am going to triage other files and open more issues. 

cc @voznesenskym if you are interested as well.

**Comment by voznesenskym** [comment]
With:
`CUDA_VISIBLE_DEVICES=-1 pytest -v -s -n 96 test/distributions/test_distributions.py`

We get:

````

FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_normal_shape_scalar_params - Asse...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_bernoulli_shape_scalar_params - A...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_chi2_shape_tensor_params - Runtim...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_binomial_shape_vectorized_n - Run...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_cauchy_shape_tensor_params - Runt...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_gamma_shape_tensor_params - Runti...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_halfcauchy_shape_tensor_params - ...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_continuous_bernoulli_shape_tensor_params
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_halfcauchy_shape_scalar_params - ...
FAILED test/distributions/test_distributions.py::TestDistributions::test_geometric - AssertionError: ValueError...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_uniform_shape_scalar_params - Ass...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_wishart_shape_tensor_params - Run...
FAILED test/distributions/test_distributions.py::TestJit::test_mean - RuntimeError: The size of tensor a (3) mu...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_binomial_shape - RuntimeError: Th...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_uniform_shape_tensor_params - Run...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_categorical_shape - RuntimeError:...
FAILED test/distributions/test_distributions.py::TestJit::test_cdf - torch.jit._trace.TracingCheckError: Tracin...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_studentT_shape_tensor_params - Ru...
FAILED test/distributions/test_distributions.py::TestJit::test_log_prob - torch.jit._trace.TracingCheckError: T...
FAILED test/distributions/test_distributions.py::TestJit::test_entropy - RuntimeError: The size of tensor a (3)...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_geometric_shape_tensor_params - R...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_laplace_shape_tensor_params - Run...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_beta_shape_scalar_params - TypeEr...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_wishart_shape_scalar_params - Typ...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_geometric_shape_scalar_params - A...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_studentT_shape_scalar_params - As...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_continuous_bernoulli_shape_scalar_params
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_bernoulli_shape_tensor_params - R...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_dirichlet_shape - RuntimeError: T...
FAILED test/distributions/test_distributions.py::TestJit::test_enumerate_support - RuntimeError: The size of te...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_normal_shape_tensor_params - Runt...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_cauchy_shape_scalar_params - Asse...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_exponential_shape_tensor_param - ...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_beta_shape_tensor_params - Runtim...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_exponential_shape_scalar_param - ...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_laplace_shape_scalar_params - Ass...
FAILED test/distributions/test_distributions.py::TestJit::test_variance - RuntimeError: The size of tensor a (3...
FAILED test/distributions/test_distributions.py::TestValidation::test_warning_unimplemented_constraints - Runti...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_multinomial_shape - RuntimeError:...
FAILED test/distributions/test_distributions.py::TestDistributionShapes::test_one_hot_categorical_shape - Asser...
FAILED test/distributions/test_distributions.py::TestValidation::test_invalid_log_probs_arg - AssertionError: S...
FAILED test/distributions/test_distributions.py::TestKL::test_kl_monte_carlo - RuntimeError: max(): Expected re...

```

I can dig into those.

**Comment by ezyang** [comment]
I have https://github.com/pytorch/pytorch/pull/79051 so we can start testing this in CI when this lands

**Comment by anijain2305** [comment]
Most of this work is in early maintenance mode now. Closing this one, will open new one for errors.
